File: java/src/main/java/com/project/conduit/service/ArticleService.java
package com.project.conduit.service;

import com.project.conduit.dto.create.ArticleDTO;
import com.project.conduit.dto.view.ArticleRO;
import com.project.conduit.dto.view.ArticlesRO;
import com.project.conduit.exception.ResourceNotFoundException;
import com.project.conduit.model.Article;
import com.project.conduit.repository.ArticleRepository;
import com.project.conduit.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.text.Normalizer;
import java.time.LocalDateTime;
import java.util.List;

@Service
public class ArticleService {
    private final ArticleRepository articleRepository;
    private final UserRepository userRepository;

    @Autowired
    public ArticleService(ArticleRepository articleRepository, UserRepository userRepository) {
        this.articleRepository = articleRepository;
        this.userRepository = userRepository;
    }

    public ArticleRO createArticle(ArticleDTO articleDTO) {
        var article = dtoToEntity(articleDTO);

        var author = userRepository.findAll().get(0);

        article.setAuthor(author);

        var savedArticle = articleRepository.save(article);

        return entityToRo(savedArticle);
    }

    public ArticlesRO findAll() {
        var articles = articleRepository.findAll();

        return entitiesToRo(articles);
    }

    public ArticleRO findBySlug(String slug) {
        var article = articleRepository.findBySlug(slug)
                .orElseThrow(() -> new ResourceNotFoundException("Article not found"));
        return entityToRo(article);
    }

    public ArticleRO updateArticle(String slug, ArticleDTO articleDTO) {
        var savedArticle = articleRepository.findBySlug(slug)
                .orElseThrow(() -> new ResourceNotFoundException("Article not found"));

        var updatedArticle = dtoToEntity(articleDTO);
        updatedArticle.setId(savedArticle.getId());

        var article = articleRepository.save(updatedArticle);

        return entityToRo(article);
    }

    public void deleteArticle(String slug) {
        var savedArticle = articleRepository.findBySlug(slug)
                .orElseThrow(() -> new ResourceNotFoundException("Article not found"));
        articleRepository.delete(savedArticle);
    }

    private Article dtoToEntity(ArticleDTO articleDTO) {
        var article = new Article();

        article.setSlug(titleToSlug(articleDTO.article().title()));
        article.setTitle(articleDTO.article().title());
        article.setDescription(articleDTO.article().description());
        article.setBody(articleDTO.article().body());
        article.setUpdatedAt(LocalDateTime.now());

        return article;
    }

    private ArticleRO entityToRo(Article article) {
        ArticleRO.AuthorDetails authorDetails = new ArticleRO.AuthorDetails(
                article.getAuthor().getUsername(),
                article.getAuthor().getBio(),
                article.getAuthor().getImage(),
                false
        );

        ArticleRO.ArticleDetails articleDetails = new ArticleRO.ArticleDetails(
                article.getSlug(),
                article.getTitle(),
                article.getDescription(),
                article.getBody(),
                article.getTagList(),
                article.getCreatedAt(),
                article.getUpdatedAt(),
                false,
                article.getFavoritesCount(),
                authorDetails
        );

        return new ArticleRO(articleDetails);
    }

    private ArticlesRO entitiesToRo(List<Article> articles) {
        List<ArticleRO.ArticleDetails> articleDetailsList = articles.stream()
                .map(article -> {
                    ArticleRO.AuthorDetails authorDetails = new ArticleRO.AuthorDetails(
                            article.getAuthor().getUsername(),
                            article.getAuthor().getBio(),
                            article.getAuthor().getImage(),
                            false
                    );

                    return new ArticleRO.ArticleDetails(
                            article.getSlug(),
                            article.getTitle(),
                            article.getDescription(),
                            article.getBody(),
                            article.getTagList(),
                            article.getCreatedAt(),
                            article.getUpdatedAt(),
                            false,
                            article.getFavoritesCount(),
                            authorDetails
                    );
                }).toList();

        return new ArticlesRO(articleDetailsList, articleDetailsList.size());
    }

    private String titleToSlug(String title) {
        if (title == null) {
            return "";
        }

        String normalized = Normalizer.normalize(title, Normalizer.Form.NFD);

        return normalized
                .replaceAll("[^\\p{Alnum} ]", "")
                .trim()
                .replaceAll("\\s+", "-")
                .toLowerCase();
    }
}

File: javascript/src/services/ArticleService.js
const articleRepository = require("../repositories/ArticleRepository");
const userRepository = require("../repositories/UserRepository");

module.exports = {
  createArticle: async (data, userId) => {
    try {
      return await articleRepository.createArticle({
        slug: titleToSlug(data.title),
        title: data.title,
        description: data.description,
        body: data.body,
        tagList: data.tagList,
        authorId: userId,
      });
    } catch (error) {
      if (error.code === "P2002") {
        const duplicateField = error.meta.target[0];
        throw new Error(
          `Duplicate ${duplicateField}: This ${duplicateField} is already taken.`
        );
      }

      throw error;
    }
  },

  findAll: async () => {
    try {
      return await articleRepository.findAll();
    } catch (error) {
      throw error;
    }
  },

  findBySlug: async (slug) => {
    try {
      return await articleRepository.findBySlug(slug);
    } catch (error) {
      throw error;
    }
  },

  updateArticle: async (id, data, userId) => {
    try {
      return await articleRepository.updateArticle(id, {
        slug: titleToSlug(data.title),
        title: data.title,
        description: data.description,
        body: data.body,
        tagList: data.tagList,
        authorId: userId,
      });
    } catch (error) {
      throw error;
    }
  },

  deleteArticle: async (id) => {
    try {
      return await articleRepository.deleteArticle(id);
    } catch (error) {
      throw error;
    }
  },
};

const titleToSlug = (title) => {
  if (title === null || title === undefined) {
    return "";
  }

  const normalized = title.normalize("NFD");

  return normalized
    .replace(/[^a-zA-Z0-9 ]/g, "")
    .trim()
    .replace(/\s+/g, "-")
    .toLowerCase();
};
